File: tests/kotlin/tests-kotlin-gradle-junit5-allure3-playwright/src/test/java/pages/LoginPage.java
package pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import io.qameta.allure.Step;

public class LoginPage {

    private final Page page;
    public final Locator loginForm;
    public final Locator loginInput;
    public final Locator passwordInput;
    public final Locator submitButton;
    public final Locator formTitle;
    public final Locator errorMessage;
    public final Locator registerLink;

    public LoginPage(Page page) {
        this.page = page;
        this.loginForm = page.getByTestId("login-form");
        this.loginInput = page.getByTestId("login-input");
        this.passwordInput = page.getByTestId("password-input");
        this.submitButton = page.getByTestId("submit-button");
        this.formTitle = page.getByTestId("login-form-title");
        this.errorMessage = page.getByTestId("error-message");
        this.registerLink = page.getByTestId("register-link");
    }

    @Step("Open login page")
    public LoginPage open() {
        page.navigate("login");
        return shouldBeOpen();
    }

    @Step("Verify login page is open")
    public LoginPage shouldBeOpen() {
        loginForm.waitFor();
        return this;
    }

    @Step("Verify login form is mounted")
    public LoginPage shouldShowLoginForm() {
        formTitle.waitFor();
        loginInput.waitFor();
        passwordInput.waitFor();
        submitButton.waitFor();
        return this;
    }

    @Step("Fill login form as {username}")
    public LoginPage login(String username, String password) {
        loginInput.fill(username);
        passwordInput.fill(password);
        submitButton.click();
        return this;
    }

    @Step("Type username: {username}")
    public LoginPage typeUsername(String username) {
        loginInput.fill(username);
        return this;
    }

    @Step("Type password")
    public LoginPage typePassword(String password) {
        passwordInput.fill(password);
        return this;
    }

    @Step("Submit login form expecting validation error")
    public LoginPage submitExpectingError() {
        submitButton.click();
        errorMessage.waitFor();
        return this;
    }

    @Step("Click Register link under the login form")
    public LoginPage clickRegisterLink() {
        registerLink.click();
        return this;
    }

    @Step("Reload current page")
    public LoginPage reload() {
        page.reload();
        return shouldBeOpen();
    }
}

File: tests/python/tests-python-playwright/pages/login_page.py
from __future__ import annotations

import allure
from playwright.sync_api import Locator, Page


class LoginPage:
    def __init__(self, page: Page) -> None:
        self.page = page
        self.login_form: Locator = page.get_by_test_id("login-form")
        self.login_input: Locator = page.get_by_test_id("login-input")
        self.password_input: Locator = page.get_by_test_id("password-input")
        self.submit_button: Locator = page.get_by_test_id("submit-button")
        self.form_title: Locator = page.get_by_test_id("login-form-title")
        self.error_message: Locator = page.get_by_test_id("error-message")
        self.register_link: Locator = page.get_by_test_id("register-link")

    @allure.step("Open login page")
    def open(self) -> LoginPage:
        self.page.goto("login")
        return self.should_be_open()

    @allure.step("Verify login page is open")
    def should_be_open(self) -> LoginPage:
        self.login_form.wait_for()
        return self

    @allure.step("Verify login form is mounted")
    def should_show_login_form(self) -> LoginPage:
        self.form_title.wait_for()
        self.login_input.wait_for()
        self.password_input.wait_for()
        self.submit_button.wait_for()
        return self

    @allure.step("Fill login form as {username}")
    def login(self, username: str, password: str) -> LoginPage:
        self.login_input.fill(username)
        self.password_input.fill(password)
        self.submit_button.click()
        return self

    @allure.step("Type username: {username}")
    def type_username(self, username: str) -> LoginPage:
        self.login_input.fill(username)
        return self

    @allure.step("Type password")
    def type_password(self, password: str) -> LoginPage:
        self.password_input.fill(password)
        return self

    @allure.step("Submit login form expecting validation error")
    def submit_expecting_error(self) -> LoginPage:
        self.submit_button.click()
        self.error_message.wait_for()
        return self

    @allure.step("Click Register link under the login form")
    def click_register_link(self) -> LoginPage:
        self.register_link.click()
        return self

    @allure.step("Reload current page")
    def reload(self) -> LoginPage:
        self.page.reload()
        return self.should_be_open()
